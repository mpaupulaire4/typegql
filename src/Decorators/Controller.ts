import { Metadata } from '../Metadata'

export function Controller() {
  return (target: Function) => {
    Metadata.controllers.push(target)
  }
}