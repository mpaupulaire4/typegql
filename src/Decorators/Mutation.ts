import { Metadata, ActionType } from '../Metadata'

interface MutationDecoratorOptions {
  name?: string
  type?: string
  args?: {
    [name: string]: string
  }
}

export function Mutation({name, type, args = {}}: MutationDecoratorOptions = {}) {
  return (target: any, methodName: string) => {
    Metadata.actions.push({
      type: ActionType.Mutation,
      target,
      methodName,
      name: name || methodName,
      static: !!target.prototype,
    })
    if (type) {
      let mutation = `${name || methodName}`
      let keys = Object.keys(args)
      if (keys.length) {
        keys = keys.map((key) => `${key}: ${args[key]}`)
        mutation = `${mutation}(\n\t\t${keys.join('\n\t\t')}\n\t)`
      }
      mutation = `${mutation}: ${type}`
      Metadata.mutations.push(mutation)
    }
  }
}

