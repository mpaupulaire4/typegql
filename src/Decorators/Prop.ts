
export const PropsMetadataKey = Symbol("types")

type PropDecoratorOptions = string

export function Prop(options?: PropDecoratorOptions) {
  return (target: any, methodName: string) => {

    const isStatic = !!target.prototype
    let constructor = isStatic ? target : target.constructor
    const types: Map<string, string> = Reflect.getMetadata(PropsMetadataKey, constructor) || new Map()
    const type = options || Reflect.getMetadata("design:type", target, methodName).name || 'NotSet'

    Reflect.defineMetadata(
      PropsMetadataKey,
      types.set(methodName, type),
      constructor
    )
  }
}


