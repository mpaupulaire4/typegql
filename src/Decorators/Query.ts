import { Metadata, ActionType } from '../Metadata'

export function Query(options: QueryDecoratorOptions = {}) {
  return (target: any, methodName: string) => {
    Metadata.actions.push({
      type: ActionType.Query,
      target,
      methodName,
      name: options.name || methodName,
      static: !!target.prototype,
    })
  }
}

interface QueryDecoratorOptions {
  name?: string
}