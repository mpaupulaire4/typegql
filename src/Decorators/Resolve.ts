
export const ResolvesMetadataKey = Symbol("resolves")

interface ResolveDecoratorOptions {
  name?: string
  dataloader?: boolean
  key?: (info: {parent: any, args: any}) => string
  type?: string
  args?: {
    [name: string]: string
  }
}

export function Resolve({ type, args = {}, ...options }: ResolveDecoratorOptions = {}) {
  return (target: any, methodName: string) => {
    const isStatic = !!target.prototype
    let constructor = isStatic ? target : target.constructor
    const resolves: Map<string, any> = Reflect.getMetadata(ResolvesMetadataKey, constructor) || new Map()
    const name = options.name || methodName
    let prop;
    if (type) {
      prop = `${name}`
      let keys = Object.keys(args)
      if (keys.length) {
        keys = keys.map((key) => `${key}: ${args[key]}`)
        prop = `${prop}(\n\t\t${keys.join('\n\t\t')}\n\t)`
      }
      prop = `${prop}: ${type}`
    }

    Reflect.defineMetadata(
      ResolvesMetadataKey,
      resolves.set(name, {
        target,
        methodName,
        name,
        dataloader: options && options.dataloader,
        key: options && options.dataloader && options.key,
        static: isStatic,
        type: prop,
      }),
      constructor
    )
  }
}


