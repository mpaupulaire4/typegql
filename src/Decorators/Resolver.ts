import { Metadata } from '../Metadata'

const TypesMetadataKey = Symbol("types")

type ResolverDecoratorOptions = string | {
  name: string
}

type ResolveDecoratorOptions = {
  name: string
  dataloader?: boolean
  key?: (info: {oarent: any, args: any}) => string
}

export function Resolver(options: ResolverDecoratorOptions) {
  return (target: any) => {
    const types: Map<string, any> = Reflect.getMetadata(TypesMetadataKey, target)
    if (!types) return
    for (let data of types.values()) {
      Metadata.resolves.push({
        ...data,
        parent: typeof options == 'string' ? options : options.name,
      })
    }
    Reflect.deleteMetadata(TypesMetadataKey, target)
  }
}

export function Resolve(options?: ResolveDecoratorOptions) {
  return (target: any, methodName: string) => {
    const isStatic = !!target.prototype
    let constructor = isStatic ? target : target.constructor
    const types: Map<string, any> = Reflect.getMetadata(TypesMetadataKey, constructor) || new Map()
    const name = options ? options.name : methodName

    Reflect.defineMetadata(
      TypesMetadataKey,
      types.set(name, {
        target,
        methodName,
        name,
        dataloader: options && options.dataloader,
        key: options && options.dataloader && options.key ,
        static: isStatic,
      }),
      constructor
    )
  }
}


