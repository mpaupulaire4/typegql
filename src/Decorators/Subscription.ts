import { Metadata, ActionType } from '../Metadata'

interface SubscriptionDecoratorOptions {
  name?: string
  listen?: string | string[],
  filter?: (payload: any, variables: any, context: any, info: any) => boolean | Promise<boolean>
  type?: string
  args?: {
    [name: string]: string
  }
}

export function Subscription({
  name,
  type,
  args = {},
  filter = () => true,
  ...options
}: SubscriptionDecoratorOptions = {}) {
  return (target: any, methodName: string) => {
    Metadata.actions.push({
      type: ActionType.Subscription,
      target,
      methodName,
      name: name || methodName,
      static: !!target.prototype,
      filter,
      listen: options.listen || name || methodName,
    })
    if (type) {
      let subscription = `${name || methodName}`
      let keys = Object.keys(args)
      if (keys.length) {
        keys = keys.map((key) => `${key}: ${args[key]}`)
        subscription = `${subscription}(\n    ${keys.join('\n    ')}\n  )`
      }
      subscription = `${subscription}: ${type}`
      Metadata.subscriptions.push(subscription)
    }
  }
}
