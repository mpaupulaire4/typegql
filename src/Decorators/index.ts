export * from './Query'
export * from './Mutation'
export * from './Subscription'
export * from './Resolve'
export * from './Type'
export * from './Prop'
