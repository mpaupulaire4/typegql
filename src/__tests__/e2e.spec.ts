// import { PubSub } from 'graphql-subscriptions';
import {
  Prop,
  Type,
  Resolve,
  Query,
  Subscription,
  Mutation,
  makeExecutableSchema
} from '../'
import { graphql } from 'graphql'
import { Container } from 'typedi'
import { PubSub } from 'graphql-subscriptions';
import { Controller } from '../Decorators';

@Type()
export class User {
  @Prop('Int') id: number;
  @Prop() name: string;
  @Prop('Int') number: number;
  constructor(
    ID: number,
    name: string,
    number: number,
  ) {
    this.id = ID
    this.name = name
    this.number = number
  }
}

let user1: User = new User(1, 'bob', 123);
let user2: User = new User(2, 'john', 456);
let user3: User = new User(3, 'jane', 789);

const bestFriendsMap =  {
  [user1.id]: user2,
  [user2.id]: user3,
  [user3.id]: user1,
}

const friendsMap =  {
  [user1.id]: [user2, user3],
  [user2.id]: [user3, user1],
  [user3.id]: [user1, user2],
}

@Type('User')
export class UserResolver {
  @Resolve({
    type: '[User!]!',
  })
  friends(user: User) {
    return friendsMap[user.id]
  }

  @Resolve({
    type: 'User',
    dataloader: true,
    key: ({ parent }) => parent.id,
  })
  bestFriend(users: User[]) {
    return users.map(user => bestFriendsMap[user.id])
  }

  @Resolve({
    type: 'User',
    dataloader: true,
    key: ({ parent }) => parent.id,
  })
  expensiveBestFriend(users: User[]) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve(users.map(user => bestFriendsMap[user.id]))
      }, 1000);
    })
  }
}

@Controller()
export class Test3 {
  @Query({
    type: '[User!]!'
  })
  users(){
    return [user1, user2, user3]
  }

  @Mutation({
    type: 'String',
  })
  mutation(){
    return ''
  }

  @Subscription({
    type: 'String'
  })
  subscription(){
    return ''
  }
}

describe('Server Setup:', () => {
  let runQuery: (query: string) => Promise<any> = () => Promise.resolve(false)

  it('should build', async () => {
    const schema = makeExecutableSchema({
      resolverBuilderOptions: {
        PubSub: new PubSub()
      }
    })
    runQuery = (query) => graphql({
      schema,
      source: query,
      contextValue: {
        container: Container
      }
    })
  })

  it('users query should return "bob", "john", and "jane"', async () => {
    const res = await runQuery(`{
      users {
        id
        name
        expensiveBestFriend {
          id
          name
          expensiveBestFriend {
            expensiveBestFriend {
              id
              name
            }
          }
        }
        bestFriend {
          id
          name
        }
        friends {
          id
          name
        }
      }
    }`)
    expect(res).toMatchSnapshot()
  })

})

