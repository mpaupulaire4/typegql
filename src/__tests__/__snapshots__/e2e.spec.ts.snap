// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server Setup: users query should return "bob", "john", and "jane" 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "bestFriend": Object {
          "name": "jane",
        },
        "friends": Array [
          Object {
            "name": "bob",
          },
          Object {
            "name": "john",
          },
        ],
        "id": 1,
        "name": "bob",
      },
      Object {
        "bestFriend": Object {
          "name": "jane",
        },
        "friends": Array [
          Object {
            "name": "bob",
          },
          Object {
            "name": "john",
          },
        ],
        "id": 2,
        "name": "john",
      },
      Object {
        "bestFriend": Object {
          "name": "jane",
        },
        "friends": Array [
          Object {
            "name": "bob",
          },
          Object {
            "name": "john",
          },
        ],
        "id": 2,
        "name": "jane",
      },
    ],
  },
}
`;
